tcp_logger_test.go

```diff
@@ -0,0 +1,63 @@
+package main
+
+import (
+	"log"
+	"net"
+	"testing"
+	"time"
+)
+
+func TestTCPLoggerBasic(t *testing.T) {
+	server, client := net.Pipe()
+
+	tcpLogger := NewTCPLogger(client, 10)
+	tcpLogger.Start()
+	tcpLogger.Write([]byte(`{"test": "testing"}\r\n`))
+
+	// TCPLogger goroutine needs a chance to process the message before test makes assertions.
+	time.Sleep(time.Millisecond * 10)
+	log.Println(len(tcpLogger.logLines))
+
+	if len(tcpLogger.logLines) > 0 {
+		t.Fatal("TCPLogger.logLines still had messages!")
+	}
+
+	tcpLogger.Close()
+	server.Close()
+}
+
+func TestTCPLoggerRetry(t *testing.T) {
+	// Server needs to accept a connnection for TCPLogger conn and then shut off
+	// to simulate bad connection.
+	server, client := net.Pipe()
+	server.Close()
+
+	tcpLogger := NewTCPLogger(client, 10)
+	// Slower sleepDuration at retryLimit in tests to ensure that the go routine
+	// that processes messages wakes up and retries the message to assert
+	// the channel successfully drains.
+	tcpLogger.sleepDuration = time.Millisecond * 10
+	tcpLogger.Start()
+	tcpLogger.Write([]byte(`{"test": "testing"}\r\n`))
+
+	// TCPLogger needs a chance for the goroutine to increment retries.
+	time.Sleep(time.Millisecond * 50)
+	if len(tcpLogger.logLines) == 0 {
+		t.Fatal("After server close, TCPLogger.logLines should not be empty if there was a failed write.")
+	}
+	if tcpLogger.retries < 1 {
+		t.Fatal("After server close, and message attempted to be written, TCPLogger.retries should >= 1")
+	}
+
+	// Set up new server/client connection and swap the connection in TCPLogger
+	server, client = net.Pipe()
+	tcpLogger.swapConn(client)
+
+	time.Sleep(time.Millisecond * 50)
+	if len(tcpLogger.logLines) > 0 {
+		t.Fatal("After client connection swap to live connection, TCPLogger.logLines should successfully push!")
```

Comment:
Is a net.Pipe blocking? i.e. do you need to spawn a go routine to read the json test payload? I'm wondering if it's stuck performing the `client.Write` call because we don't call `server.Read`. Might be worth ensuring the message that comes out is `{$test_json|}\r\n`.

Reply:
This was a good catch.  It definitely was hanging during the `client.Write`. I added a routine to close the writer just before blocking a read in the test, which now asserts message equivalence.